packages/flutter_driver/lib/src/driver/vmservice_driver.dart

```diff
@@ -96,11 +96,23 @@ class VMServiceFlutterDriver extends FlutterDriver {
     final VMServiceClientConnection connection =
     await vmServiceConnectFunction(dartVmServiceUrl, headers: headers);
     final VMServiceClient client = connection.client;
-    final VM vm = await client.getVM();
-    final VMIsolateRef isolateRef = isolateNumber ==
-        null ? vm.isolates.first :
-    vm.isolates.firstWhere(
-            (VMIsolateRef isolate) => isolate.number == isolateNumber);
+
+    VMIsolateRef isolateRef;
+    const int totalTries = 10;
+    for (int tries = 0; tries < totalTries; tries += 1) {
+      final VM vm = await client.getVM();
+      if (vm.isolates.isEmpty) {
+        await Future<void>.delayed(_kPauseBetweenReconnectAttempts);
+        continue;
+      }
+      isolateRef = isolateNumber == null
+        ? vm.isolates.first
+        : vm.isolates.firstWhere((VMIsolateRef isolate) => isolate.number == isolateNumber);
+      break;
+    }
+    if (isolateRef == null) {
+      throw DriverError('Failed to find main isolate after ${_kPauseBetweenReconnectAttempts.inSeconds * totalTries} seconds!');
```

Comment:
In other places, like `vmServiceConnectFunction` (which is `_waitAndConnec`), this code doesn't throw an exception for a connection timeout. Instead it emits a log that the operation is taking a long time. Is there some reason not to follow that convention here?

Reply:
I could treat it that way here too, which would then just defer to the global test timeout. Will update.

Reply:
Done.